lib/remap-lang/src/type_def.rs

```diff
@@ -206,6 +206,20 @@ impl TypeDef {
         self
     }
 
+    /// Applies a type constraint to the items in an array. If you need all items in the array to
+    /// be integers, for example, set `Kind::Integer`; if items can be either integers or Booleans,
+    /// set `Kind::Integer | Kind::Boolean`; and so on.
+    pub fn fallible_unless_array_has_inner_type(mut self, kind: impl Into<value::Kind>) -> Self {
```

Comment:
:+1: to this constraint, but the naming / implementation feels a bit awkward to me. This might be due to the fact that `inner_type` is a bit awkward right now in-general, but, based on the current naming, I'm wondering if this function shouldn't be setting `self.fallible = true` if a non-array type is passed into it.

Another suggestion would be to call this `fallible_unless_inner_type(...)` and implement it such that it could accept any of our types that has an inner type.

In my dream world, I think maybe it'd make sense to have `Kind::Array` itself be an enum type so that you could do things like: `Kind::Array(Kind::Integer | Kind::Boolean)` to represent an array of integers/booleans, but that would be a more substantive refactoring.

Reply:
@jszwedko Good call on that. I've updated the behavior to set `fallible` to `true` for non-array types (plus a test for that 😎). I do like your suggestion about the desired end state but I'll leave that to one of our Remap Hero(in)es.

Reply:
This function could be made much more generic. Some of the different rules we could want:

- a map that has a certain field of a certain type
- an array of maps that has a certain field of a certain type
- either an array of strings or an array of maps that have a certain field of a certain type (say, either `["a", "b", "c"]` or `[{message: "a"}, {message: "b"}, {message: "c"}]`

I think we should wait to see if we actually ever need that level of flexibility before tackling it because it would take a fair bit of thought to get right. So for now I'd say it's good enough to have a fairly specific function that does what we need, but no more. It does result in the awkward name, but I think that reflects the specific nature of where we are up to so far!

> In my dream world.

I share this vision! The issue here is that `Kind` is a bit flag, so `Kind::Array | Kind::String` is still a `Kind`. Which is nice. If we make `Kind:Array` an an enum type, it could no longer be a bit flag. We would then need to distinguish between primitive Kinds and complex Kinds.

Reply:
> I share this vision! The issue here is that `Kind` is a bit flag, so `Kind::Array | Kind::String` is still a `Kind`. Which is nice. If we make `Kind:Array` an an enum type, it could no longer be a bit flag. We would then need to distinguish between primitive Kinds and complex Kinds.

Agreed, that seems to be the distinction here :+1: